src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -3533,7 +3535,7 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
                   (argTyped, argTyped.tpe.deconst)
                 }
 
-                arg match {
+                val (arg1,argTpe1) = arg match {
```

Comment:
**Question**: We can remove this tuple by simply putting the list insertions in each case. We can also remove the tuple created in the `typedArg0` inner function above, by adding the `.tpe.deconst` in each case. 

Before doing that change, I would like to ask: can the compiler optimise those temporary tuples away?

Reply:
The optimizing backend does some tuple elimination, but across method boundaries this only works if the method is inlined.

Reply:
Ok. I ended up removing the tuples manually. It does not add too many lines.

Reply:
I've pushed a further commit with the remainder of my review suggestions :)